core/src/main/scala/kafka/zk/ZkData.scala

```diff
@@ -30,6 +30,13 @@ import org.apache.zookeeper.data.Stat
 
 // This file contains objects for encoding/decoding data stored in ZooKeeper nodes (znodes).
 
+object LeaderAndIsrZNode {
+  def encode(leaderAndIsr: LeaderAndIsr, controllerEpoch: Int): String = {
```

Comment:
Other `encode` methods return `Array[Byte]`, why does this return a `String`?

Reply:
This method calls Json.encode().
The String return type was from zkUtils.leaderAndIsrZkData().

Reply:
The point is that we should aim to be consistent and return `Array[Byte]` from all the `encode` methods.

Reply:
String is expected for argument 2 of:
```
  def conditionalUpdatePath(path: String, data: String, expectVersion: Int,
```
If return value is changed to Array[Byte], wouldn't that require converting the output from Json.encode() to Array[Byte] and back to String ?
Please confirm.

Reply:
You should change `data` in that method to be of type `Array[Byte]`.

Reply:
The parameter meanders thru checker and lands as 1st parameter here:
```
  private def parseLeaderAndIsr(leaderAndIsrStr: String, path: String, stat: Stat): Option[LeaderIsrAndControllerEpoch] = {
    Json.parseFull(leaderAndIsrStr).flatMap { js =>
```
This would involve back and forth conversion which seems unnecessary.

Reply:
That code can be changed to be `Json.parseBytes`. You can take a look at the other code, it's all following the same pattern.

Reply:
When making the change, I found I need to overcome the following:

https://pastebin.com/rphaH3eD

Can you confirm this is desirable ?

Reply:
I think it's desirable, but maybe we can leave it for another PR in that case.

Reply:
Yeah. The scope is bigger than I thought.